File: app/frontend/src/markdown.ts
import { marked, type Token } from 'marked'

// ---------------------------------------------------------------------------
// Shared markdown rendering — no wasm dependency, used by both the live app
// (via markdown-live.ts's renderMarkdown) and the SPA export
// (renderMarkdownFromEntries below). Keeping this module wasm-free lets the
// export's Vite entry point skip bundling the wasm splitter entirely, which
// matters because the export can be opened via `file://` (no fetch of the
// .wasm asset works there) — see markdown-live.ts for the wasm-backed path.
// ---------------------------------------------------------------------------

// ---------------------------------------------------------------------------
// Silent text — bracketed spans followed by a `<!--silent-->` comment, e.g.
// `[Doug Engelbart]<!--silent-->`. Displayed (brackets kept) but excluded from
// TTS and playback. See dev/silent-text.md. Mirrors strip_silent in
// tts/src/markdown.rs.
// ---------------------------------------------------------------------------

const SILENT_SPAN = /\[[^\]]*\]\s*<!--\s*silent\s*-->/g

// True when a whole block (heading/paragraph) is nothing but a single silent
// span — the only case handled in this first pass (mid-sentence inline silent
// is deferred).
const FULLY_SILENT = /^\s*\[[^\]]*\]\s*<!--\s*silent\s*-->\s*$/

function isFullySilent(text: string): boolean {
  return FULLY_SILENT.test(text)
}

// The bracketed display text for a silent span, with the comment stripped and
// the brackets kept (the editorial-insertion convention).
function silentDisplayText(text: string): string {
  return text.replace(/<!--\s*silent\s*-->/g, '').trim()
}

// Outline label for a heading: bracketed display text if silent, otherwise
// the inline-stripped plain text.
function silentOrPlain(text: string): string {
  return isFullySilent(text) ? silentDisplayText(text) : stripInline(text)
}

// Remove silent spans to derive the plain text fed to TTS. Drops a line that
// became empty (or only heading `#` markers) because of the stripping;
// preserves originally-blank lines so paragraph boundaries survive.
export function stripSilent(markdown: string): string {
  const out: string[] = []
  for (const line of markdown.split('\n')) {
    const removed = line.replace(SILENT_SPAN, '')
    const trimmed = removed.trim()
    if (removed !== line && (trimmed === '' || /^#+$/.test(trimmed))) continue
    out.push(removed)
  }
  return out.join('\n')
}

// Strip inline markdown markers to get plain text. Only needs to handle what
// trafilatura produces: bold, italic, code, links. Exported for
// markdown-live.ts's wasm-backed sentence provider.
export function stripInline(text: string): string {
  return text
    .replace(/\*\*(.*?)\*\*/gs, '$1')
    .replace(/__(.*?)__/gs, '$1')
    .replace(/\*(.*?)\*/gs, '$1')
    .replace(/_(.*?)_/gs, '$1')
    .replace(/`(.*?)`/g, '$1')
    .replace(/\[([^\]]+)\]\([^)]+\)/g, '$1')
}

// ---------------------------------------------------------------------------
// renderMarkdownFromEntries — export-only entry point. (Live app's
// renderMarkdown lives in markdown-live.ts.)
//
// Parses `content` (trafilatura markdown) and appends rendered HTML into
// `container`. Sentence spans are woven into each block element so the
// Player can activate them in place as audio arrives.
//
// Returns `pendingSpans` in synthesis order, and `headings` for the outline.
// ---------------------------------------------------------------------------

export interface HeadingEntry {
  depth: number
  text: string          // plain text, for display in the outline
  element: HTMLElement  // the hN element in the DOM — scroll target
  sentenceIndex: number // global sentence index of the heading's first sentence
}

export interface RenderResult {
  pendingSpans: HTMLElement[]
  headings: HeadingEntry[]
}

// The SPA export already knows exact sentence boundaries and per-sentence
// markdown text, computed once at export time by `tts::markdown::split_for_export`
// (no wasm needed). `blockLengths` gives the sentence count for each block in
// the same document-walk order `marked.lexer` produces, so blocks can be
// matched up positionally without re-splitting anything client-side.
export interface ExportSentenceEntry {
  text: string
  markdown_text: string
}

export function renderMarkdownFromEntries(
  content: string,
  entries: ExportSentenceEntry[],
  blockLengths: number[],
  container: HTMLElement,
): RenderResult {
  return renderWithProvider(content, entriesSentenceProvider(entries, blockLengths), container)
}

// Used by markdown-live.ts's renderMarkdown for the live app's wasm-backed
// path — keeps the marked.lexer block walk in one place.
export function renderWithProvider(
  content: string,
  provider: SentenceProvider,
  container: HTMLElement,
): RenderResult {
  const pendingSpans: HTMLElement[] = []
  const headings: HeadingEntry[] = []
  let globalIdx = 0
  const fragment = document.createDocumentFragment()
  const tokens = marked.lexer(content)
  for (const token of tokens) {
    globalIdx = renderToken(token, fragment, provider, globalIdx, pendingSpans, headings)
  }
  container.appendChild(fragment)
  return { pendingSpans, headings }
}

// ---------------------------------------------------------------------------
// Block rendering
// ---------------------------------------------------------------------------

// Returns the updated globalIdx after consuming sentences for this token.
function renderToken(
  token: Token,
  container: HTMLElement | DocumentFragment,
  provider: SentenceProvider,
  globalIdx: number,
  pendingSpans: HTMLElement[],
  headings: HeadingEntry[],
): number {
  switch (token.type) {
    case 'heading': {
      const el = document.createElement(`h${token.depth}`)
      el.className = 'md-heading'
      const sentenceIndex = globalIdx
      if (isFullySilent(token.text)) {
        // Display-only navigation heading: shown in body + outline, never
        // spoken. No span woven, globalIdx unchanged, so it points at the
        // next real sentence — the natural scroll target.
        el.classList.add('silent')
        el.textContent = silentDisplayText(token.text)
      } else {
        globalIdx = provider.weave(token.text, el, globalIdx, pendingSpans)
      }
      container.appendChild(el)
      headings.push({ depth: token.depth, text: silentOrPlain(token.text), element: el, sentenceIndex })
      break
    }
    case 'paragraph': {
      const el = document.createElement('p')
      el.className = 'md-paragraph'
      if (isFullySilent(token.text)) {
        el.classList.add('silent')
        el.textContent = silentDisplayText(token.text)
      } else {
        globalIdx = provider.weave(token.text, el, globalIdx, pendingSpans)
      }
      container.appendChild(el)
      break
    }
    case 'blockquote': {
      const el = document.createElement('blockquote')
      el.className = 'md-blockquote'
      for (const child of (token as any).tokens ?? []) {
        globalIdx = renderToken(child, el, provider, globalIdx, pendingSpans, headings)
      }
      container.appendChild(el)
      break
    }
    case 'list': {
      const el = document.createElement(token.ordered ? 'ol' : 'ul')
      el.className = 'md-list'
      for (const item of token.items) {
        const li = document.createElement('li')
        li.className = 'md-list-item'
        globalIdx = provider.weave(item.text, li, globalIdx, pendingSpans)
        el.appendChild(li)
      }
      container.appendChild(el)
      break
    }
    case 'code': {
      const pre = document.createElement('pre')
      pre.className = 'md-code'
      const code = document.createElement('code')
      code.textContent = token.text
      pre.appendChild(code)
      container.appendChild(pre)
      break
    }
    case 'hr': {
      container.appendChild(document.createElement('hr'))
      break
    }
    case 'space':
      break
    default: {
      const text = (token as any).text as string | undefined
      if (text?.trim()) {
        const el = document.createElement('p')
        el.className = 'md-paragraph'
        globalIdx = provider.weave(text, el, globalIdx, pendingSpans)
        container.appendChild(el)
      }
    }
  }
  return globalIdx
}

// ---------------------------------------------------------------------------
// Sentence span weaving
//
// renderToken doesn't split text into sentences itself — it delegates to a
// SentenceProvider, since the live app and the SPA export get their
// per-sentence text from different places (wasm split at render time vs.
// a precomputed export payload). Both providers append the same `.segment`
// spans in the same shape; only how they find sentence text differs.
// ---------------------------------------------------------------------------

export interface SentenceProvider {
  // Weaves spans for one block's sentences into `container`, returns the
  // updated globalIdx.
  weave(rawText: string, container: HTMLElement, globalIdx: number, pendingSpans: HTMLElement[]): number
}

// SPA export: sentence counts and markdown text were already computed at
// export time (see `tts::markdown::split_for_export`), so this provider
// only needs to consume `blockLengths`/`entries` in document order — no
// splitting, no wasm.
function entriesSentenceProvider(entries: ExportSentenceEntry[], blockLengths: number[]): SentenceProvider {
  const lengthsQueue = [...blockLengths]
  return {
    weave(_rawText, container, globalIdx, pendingSpans) {
      const count = lengthsQueue.shift() ?? 0
      for (let i = 0; i < count; i++) {
        const entry = entries[globalIdx]
        const span = document.createElement('span')
        span.className = 'segment pending'
        span.innerHTML = marked.parseInline(entry?.markdown_text ?? '') as string
        pendingSpans.push(span)
        container.appendChild(span)
        if (i < count - 1) {
          container.appendChild(document.createTextNode(' '))
        }
        globalIdx++
      }
      return globalIdx
    },
  }
}

File: tts/src/markdown.rs
//! Markdown -> plain text conversion, for TTS sentence splitting.
//!
//! Mirrors the inline-stripping behavior of the frontend's `markdown.ts`:
//! bold/italic/inline-code markers are removed and links are reduced to
//! their link text. Fenced code blocks, images, and horizontal rules are
//! dropped entirely (matching `markdown.ts`, which never weaves code-block
//! content into the spoken sentence list). Block-level elements (paragraphs,
//! headings, list items, blockquote text) are joined with blank lines so
//! the result can be split into paragraphs the same way as plain-text input.

use pulldown_cmark::{Event, Parser, Tag, TagEnd};

/// One spoken sentence, with both its plain text (for synthesis indices)
/// and its raw markdown source (for inline-formatted export rendering).
#[derive(Debug, Clone, PartialEq)]
pub struct ExportSentence {
    pub text: String,
    pub markdown_text: String,
    pub paragraph_end: bool,
}

/// Split `markdown` into sentences for static-export rendering, paired with
/// per-block sentence counts so a caller walking the same block structure
/// (e.g. the frontend's `marked.lexer`) knows how many sentences each block
/// contributes — mirrors `markdown.ts`'s `weaveSpans`, but run once at
/// export time instead of per-render in the browser, so the export never
/// needs the wasm splitter.
///
/// Per block, the plain-text split (via `splitter::split`) determines the
/// sentence count; the raw-markdown split is used for the rendered text only
/// when its count matches — otherwise (same fallback `weaveSpans` uses) the
/// plain text is repeated as `markdown_text` for that block, since a count
/// mismatch means the raw split can't be trusted to align sentence-for-
/// sentence with the plain one.
pub fn split_for_export(markdown: &str) -> (Vec<ExportSentence>, Vec<usize>) {
    let markdown = strip_silent(markdown);

    let mut sentences: Vec<ExportSentence> = Vec::new();
    let mut block_lengths: Vec<usize> = Vec::new();

    let mut current_plain = String::new();
    let mut block_start: Option<usize> = None;
    let mut last_end = 0usize;
    let mut in_code_block = false;
    let mut in_image = false;

    for (event, range) in Parser::new(&markdown).into_offset_iter() {
        match event {
            Event::Start(Tag::CodeBlock(_)) => in_code_block = true,
            Event::End(TagEnd::CodeBlock) => in_code_block = false,
            Event::Start(Tag::Image { .. }) => in_image = true,
            Event::End(TagEnd::Image) => in_image = false,
            Event::Start(Tag::Paragraph | Tag::Heading { .. } | Tag::Item) => {
                block_start = Some(range.start);
            }
            Event::Text(text) | Event::Code(text) => {
                if !in_code_block && !in_image {
                    current_plain.push_str(&text);
                }
            }
            Event::SoftBreak | Event::HardBreak => {
                current_plain.push(' ');
            }
            Event::End(TagEnd::Paragraph | TagEnd::Heading(_) | TagEnd::Item) => {
                if let Some(start) = block_start.take() {
                    let raw = markdown[start..last_end].trim();
                    push_export_block(&mut sentences, &mut block_lengths, &current_plain, raw);
                }
                current_plain.clear();
            }
            _ => {}
        }
        last_end = range.end;
    }

    (sentences, block_lengths)
}

/// Collapse a markdown hard break (line ending in 2+ spaces, or a trailing
/// backslash) to a literal `<br>` so it survives sentence splitting and can
/// still be rendered as a line break; a soft break collapses to a space.
/// Mirrors `collapseHardBreaksToBr` in `markdown.ts`.
fn collapse_hard_breaks(text: &str) -> String {
    let mut out = String::with_capacity(text.len());
    let mut rest = text;
    while let Some(idx) = rest.find('\n') {
        let before = &rest[..idx];
        if before.ends_with("  ") {
            out.push_str(before.trim_end_matches(' '));
            out.push_str(" <br>");
        } else if before.ends_with('\\') {
            out.push_str(&before[..before.len() - 1]);
            out.push_str("<br>");
        } else {
            out.push_str(before);
            out.push(' ');
        }
        rest = &rest[idx + 1..];
    }
    out.push_str(rest);
    out
}

fn push_export_block(
    sentences: &mut Vec<ExportSentence>,
    block_lengths: &mut Vec<usize>,
    plain: &str,
    raw: &str,
) {
    let plain = plain.split_whitespace().collect::<Vec<_>>().join(" ");
    if plain.is_empty() {
        return;
    }
    let plain_sentences = crate::splitter::split(&plain);
    let raw_collapsed = collapse_hard_breaks(raw);
    let raw_sentences = crate::splitter::split(&raw_collapsed);

    let use_raw = raw_sentences.len() == plain_sentences.len();
    block_lengths.push(plain_sentences.len());
    for (i, s) in plain_sentences.into_iter().enumerate() {
        let markdown_text = if use_raw {
            raw_sentences[i].text.clone()
        } else {
            s.text.clone()
        };
        sentences.push(ExportSentence {
            text: s.text,
            markdown_text,
            paragraph_end: s.paragraph_end,
        });
    }
}

/// Convert markdown to plain text suitable for sentence splitting and TTS.
pub fn to_plain_text(markdown: &str) -> String {
    let markdown = strip_silent(markdown);

    let mut blocks: Vec<String> = Vec::new();
    let mut current = String::new();
    let mut in_code_block = false;
    let mut in_image = false;

    for event in Parser::new(&markdown) {
        match event {
            Event::Start(Tag::CodeBlock(_)) => in_code_block = true,
            Event::End(TagEnd::CodeBlock) => in_code_block = false,
            Event::Start(Tag::Image { .. }) => in_image = true,
            Event::End(TagEnd::Image) => in_image = false,
            Event::Text(text) | Event::Code(text) => {
                if !in_code_block && !in_image {
                    current.push_str(&text);
                }
            }
            Event::SoftBreak | Event::HardBreak => {
                current.push(' ');
            }
            Event::End(
                TagEnd::Paragraph
                | TagEnd::Heading(_)
                | TagEnd::Item
                | TagEnd::TableCell,
            ) => {
                push_block(&mut blocks, &current);
                current.clear();
            }
            _ => {}
        }
    }

    push_block(&mut blocks, &current);

    blocks.join("\n\n")
}

/// Remove "silent" spans — bracketed text immediately followed by a
/// `<!--silent-->` comment, e.g. `[Doug Engelbart]<!--silent-->`. These are
/// displayed in the rendered document but excluded from speech synthesis.
/// See `dev/silent-text.md`.
///
/// Operates line by line: silent spans are stripped from each line, and a
/// line that became empty (or only heading `#` markers) *because* of the
/// stripping is dropped entirely — so a fully-silent heading produces no
/// spoken block. Originally-blank lines are preserved so paragraph
/// boundaries survive.
fn strip_silent(markdown: &str) -> String {
    let mut out: Vec<String> = Vec::new();
    for line in markdown.lines() {
        let removed = remove_silent_spans(line);
        let trimmed = removed.trim();
        let emptied_by_strip = removed != line
            && (trimmed.is_empty() || trimmed.chars().all(|c| c == '#'));
        if emptied_by_strip {
            continue;
        }
        out.push(removed);
    }
    out.join("\n")
}

/// Remove every `[...]<!--silent-->` span from a single line, keeping any
/// other text. A `<!--silent-->` comment not preceded by a bracket span has
/// only the comment removed.
fn remove_silent_spans(line: &str) -> String {
    let mut out = String::new();
    let mut rest = line;
    while let Some((start, end)) = find_silent_comment(rest) {
        let before = &rest[..start];
        if before.ends_with(']') {
            if let Some(open) = before.rfind('[') {
                out.push_str(&before[..open]);
                rest = &rest[end..];
                continue;
            }
        }
        out.push_str(before);
        rest = &rest[end..];
    }
    out.push_str(rest);
    out
}

/// Find the next `<!--silent-->` comment (whitespace around `silent`
/// tolerated), returning its byte range `(start, end)` within `s`.
fn find_silent_comment(s: &str) -> Option<(usize, usize)> {
    let mut from = 0;
    while let Some(rel) = s[from..].find("<!--") {
        let start = from + rel;
        let inner_start = start + 4;
        match s[inner_start..].find("-->") {
            Some(erel) => {
                let inner_end = inner_start + erel;
                let end = inner_end + 3;
                if s[inner_start..inner_end].trim() == "silent" {
                    return Some((start, end));
                }
                from = end;
            }
            None => return None,
        }
    }
    None
}

/// Trim and collapse internal whitespace runs (left by dropped inline
/// elements like images) before pushing a finished block.
fn push_block(blocks: &mut Vec<String>, text: &str) {
    let collapsed: String = text.split_whitespace().collect::<Vec<_>>().join(" ");
    if !collapsed.is_empty() {
        blocks.push(collapsed);
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn strips_bold_italic_inline_code() {
        let md = "This is **bold**, *italic*, and `code`.";
        assert_eq!(to_plain_text(md), "This is bold, italic, and code.");
    }

    #[test]
    fn reduces_links_to_text() {
        let md = "See [this reference](https://example.com) for details.";
        assert_eq!(to_plain_text(md), "See this reference for details.");
    }

    #[test]
    fn joins_paragraphs_with_blank_line() {
        let md = "First paragraph.\n\nSecond paragraph.";
        assert_eq!(to_plain_text(md), "First paragraph.\n\nSecond paragraph.");
    }

    #[test]
    fn includes_headings_and_list_items() {
        let md = "# Title\n\n- one\n- two";
        assert_eq!(to_plain_text(md), "Title\n\none\n\ntwo");
    }

    #[test]
    fn skips_fenced_code_blocks() {
        let md = "Before.\n\n```\nlet x = 1;\n```\n\nAfter.";
        assert_eq!(to_plain_text(md), "Before.\n\nAfter.");
    }

    #[test]
    fn skips_images() {
        let md = "Look: ![alt text](pic.png) done.";
        assert_eq!(to_plain_text(md), "Look: done.");
    }

    #[test]
    fn drops_silent_heading() {
        let md = "## [Doug Engelbart]<!--silent-->\n\nHe invented the mouse.";
        assert_eq!(to_plain_text(md), "He invented the mouse.");
    }

    #[test]
    fn drops_standalone_silent_paragraph() {
        let md = "First.\n\n[An aside]<!--silent-->\n\nSecond.";
        assert_eq!(to_plain_text(md), "First.\n\nSecond.");
    }

    #[test]
    fn preserves_paragraph_boundaries_around_silent() {
        // The blank lines separating paragraphs must survive stripping.
        let md = "# [Intro]<!--silent-->\n\nFirst para.\n\nSecond para.";
        assert_eq!(to_plain_text(md), "First para.\n\nSecond para.");
    }

    #[test]
    fn tolerates_whitespace_in_marker() {
        let md = "## [Ted Nelson]<!-- silent -->\n\nHe coined hypertext.";
        assert_eq!(to_plain_text(md), "He coined hypertext.");
    }

    // ── split_for_export ──────────────────────────────────────────────────

    #[test]
    fn export_single_sentence_block_keeps_formatting() {
        let md = "This is **bold** text.";
        let (sentences, block_lengths) = split_for_export(md);
        assert_eq!(block_lengths, vec![1]);
        assert_eq!(sentences[0].text, "This is bold text.");
        assert_eq!(sentences[0].markdown_text, "This is **bold** text.");
    }

    #[test]
    fn export_multiple_sentences_in_one_block() {
        let md = "First sentence. Second **sentence**.";
        let (sentences, block_lengths) = split_for_export(md);
        assert_eq!(block_lengths, vec![2]);
        assert_eq!(sentences[0].text, "First sentence.");
        assert_eq!(sentences[0].markdown_text, "First sentence.");
        assert_eq!(sentences[1].text, "Second sentence.");
        assert_eq!(sentences[1].markdown_text, "Second **sentence**.");
    }

    #[test]
    fn export_heading_and_list_items_are_separate_blocks() {
        let md = "# Title\n\n- one\n- two";
        let (sentences, block_lengths) = split_for_export(md);
        assert_eq!(block_lengths, vec![1, 1, 1]);
        assert_eq!(sentences.iter().map(|s| s.text.as_str()).collect::<Vec<_>>(), vec!["Title", "one", "two"]);
    }

    #[test]
    fn export_paragraph_with_link_preserves_markdown() {
        let md = "See [this reference](https://example.com) for details.";
        let (sentences, block_lengths) = split_for_export(md);
        assert_eq!(block_lengths, vec![1]);
        assert_eq!(sentences[0].text, "See this reference for details.");
        assert_eq!(sentences[0].markdown_text, "See [this reference](https://example.com) for details.");
    }

    #[test]
    fn export_drops_silent_heading_block() {
        let md = "## [Doug Engelbart]<!--silent-->\n\nHe invented the mouse.";
        let (sentences, block_lengths) = split_for_export(md);
        assert_eq!(block_lengths, vec![1]);
        assert_eq!(sentences[0].text, "He invented the mouse.");
    }
}
